File: CrackingTheCodingInterview/ch04-TreesAndGraphs/randomNode.js
/*
  in addition to insert, find, and delete methods for a binary
  tree class, implement a getRandomNode() method that returns
  a random node from the tree
*/

function BinaryTree(val) {
  this.value = val;
  this.size = 1;
}

BinaryTree.prototype.insert = (val) => {
  if (val < this.value) {
    if (!this.left) {
      this.left = new BinaryTree(val);
    } else {
      this.left.insert(val);
    }
  } else {
    if (!this.right) {
      this.right = new BinaryTree(val);
    } else {
      this.right.insert(val);
    }
  }
  this.size++;
}

// each node should have 1/size chance of getting picked
BinaryTree.prototype.getRandomNode = () => {
  // breakpoints
  // left: < left.size
  // right: >= left.size and < left.size + right.size
  // current node: 1
  const random = Math.random() * this.size;
  if (random < this.left.size) {
    return this.left.getRandomNode();
  } else if (random >= this.left.size &&
    random < this.left.size + this.right.size) {
    return this.right.getRandomNode();
  } else {
    return this;
  }
}

File: CrackingTheCodingInterview/ch04-TreesAndGraphs/randomNode.py
"""
In addition to insert, find, and delete methods for a binary
tree class, implement a get_random_node() method that returns
a random node from the tree
"""
from random import randrange


class BSTNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        self.size = 1

    def insert(self, value):
        # If value already in tree
        if value == self.value:
            return

        # Recursively insert
        if value < self.value:
            if not self.left:
                self.left = BSTNode(value)
            else:
                self.left.insert(value)
        else:
            if not self.right:
                self.right = BSTNode(value)
            else:
                self.right.insert(value)

        # Bubble up size increment
        self.size += 1

    # Each node should have a 1/size chance of getting picked
    # |---------------|-------------------------------|-|
    #   <left.size      >=left.size and <size-1        ==size-1
    def get_random_node(self):
        r = randrange(self.size)

        if r < self.left.size:
            return self.left.get_random_node()
        elif self.left.size <= r < self.size - 1:
            return self.right.get_random_node()
        else:
            return self
